### PR TITLE
[libc++][regex] add _LIBCPP_FALLTHROUGH to suppress fallthrough warning

### PR DESCRIPTION
**Pull request by vegerot**
Summary:
The diff https://github.com/llvm/llvm-project/pull/97926 is stacked on top of this patch because this file reports an error when enabling `-Wimplicit-fallthrough` in `-Wextra`.

Test plan:

```sh
$ time (mkdir build_runtimes && cd build_runtimes && set -x && CC=../build/bin/clang CXX=../build/bin/clang++ cmake -G Ninja ../runtimes -DCMAKE_BUILD_TYPE=Release -DLLVM_ENABLE_RUNTIMES='libcxx;libcxxabi;libunwind'  && ninja && bin/llvm-lit -sv ../libcxx/test/std/re )
```

note: whether I put a `break;` or fallthrough, the tests pass anyways which is sus.

### CODE REVIEW

File: libcxx/include/regex
Comment (vegerot) [inline libcxx/include/regex:3924]: Question for reviewer: this fallthrough looks strange.  Are we sure it shouldn't be a `break;`?

I should probably change this to a `break;` and see what tests break...

**Comment by vegerot** [comment]
question for reviewer: what is the proper etiquette for stacked diffs? Should I open two PRs or have one PR with two commits?

File: libcxx/include/regex
Comment (frederick-vs-ja) [reply libcxx/include/regex:3924]: It seems to me that we should spell the attribute as `[[_Clang::__fallthrough__]]` to defend against user-defined macros. (As `clang` is never reserved and `fallthrough` is reserved only since C++17.)
Or can we use `[[__fallthrough__]]`?

Comment (philnik777) [reply libcxx/include/regex:3924]: Given that there was a `fallthrough` comment, I'd assume it was intentional. And, as @frederick-vs-ja we should make this `[[__fallthrough__]]` instead.

**Comment by mordante** [comment]
> question for reviewer: what is the proper etiquette for stacked diffs? Should I open two PRs or have one PR with two commits?

It's only possible to make stacked reviews for members of the LLVM project since it requires write access to the repository. I prefer you wait uploading the next patch until this one lands. Having two patches mixed causes reviewers to review the same code twice.

File: libcxx/include/regex
Comment (vegerot) [reply libcxx/include/regex]: 1. done
2. Whether I put a `break;` or fallthrough here, the tests pass which is a little sus. (thanks to @philnik777 for [helping](https://discord.com/channels/636084430946959380/636732894974312448/1267222316967002244) me run the tests)

File: libcxx/include/regex
Comment (ldionne) [inline libcxx/include/regex:3924]: The CI failure is caused by AppleClang not accepting this syntax of attributes in C++03 mode. Can you use `_LIBCPP_FALLTHROUGH` instead?

Comment (vegerot) [reply libcxx/include/regex:3924]: This doesn't work for me.  I made this change and ran 

```sh
time (mkdir build_runtimes && cd build_runtimes && set -x && CC=../build/bin/clang CXX=../build/bin/clang++ cmake -G Ninja ../runtimes -DCMAKE_BUILD_TYPE=Release -DLLVM_ENABLE_RUNTIMES='libcxx;libcxxabi;libunwind'  && ninja && bin/llvm-lit -sv ../libcxx/test/std/re )
```

and it errored with

```cpp
/home/max/workspace/github.com/llvm/llvm-project/build_runtimes/include/c++/v1/regex:3924:9: warning: unknown attribute '_LIBCPP_FALLTHROUGH' ignored [-Wunknown-attributes]
 3924 |       [[_LIBCPP_FALLTHROUGH]];
      |         ^~~~~~~~~~~~~~~~~~~
1 warning generated.
```

Comment (frederick-vs-ja) [reply libcxx/include/regex:3924]: Existing code uses `_LIBCPP_FALLTHROUGH();`.
```suggestion
      _LIBCPP_FALLTHROUGH();
```

**Review by ldionne — APPROVED** [review]
LGTM, this can be merged once CI is green.

**Pull request merged**
